targets/simple_switch_grpc/tests/test_grpc_counter.cpp

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2017, Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <grpc++/grpc++.h>
+
+#include <p4/bm/dataplane_interface.grpc.pb.h>
+#include <p4/p4runtime.grpc.pb.h>
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <string>
+
+#include "base_test.h"
+
+namespace sswitch_grpc {
+
+namespace testing {
+
+namespace {
+
+constexpr char counter_json[] = TESTDATADIR "/counter.json";
+constexpr char counter_proto[] = TESTDATADIR "/counter.proto.txt";
+
+class SimpleSwitchGrpcTest_Counter : public SimpleSwitchGrpcBaseTest {
+ protected:
+  SimpleSwitchGrpcTest_Counter()
+      : SimpleSwitchGrpcBaseTest(counter_proto),
+        dataplane_channel(grpc::CreateChannel(
+            dp_grpc_server_addr, grpc::InsecureChannelCredentials())),
+        dataplane_stub(p4::bm::DataplaneInterface::NewStub(
+            dataplane_channel)) { }
+
+  void SetUp() override {
+    SimpleSwitchGrpcBaseTest::SetUp();
+    update_json(counter_json);
+  }
+
+  std::shared_ptr<grpc::Channel> dataplane_channel{nullptr};
+  std::unique_ptr<p4::bm::DataplaneInterface::Stub> dataplane_stub{nullptr};
+};
+
+TEST_F(SimpleSwitchGrpcTest_Counter, CounterHit) {
+  auto t_id = get_table_id(p4info, "t_redirect");
+  auto mf_id = get_mf_id(p4info, "t_redirect", "sm.packet_length");
+  auto a_id = get_action_id(p4info, "port_redirect");
+  auto dc_id = get_dc_id(p4info, "cntr");
+
+  p4::Entity entity;
```

Comment:
based on how the test is organized, I think it would look better as a helper function:
```
p4::WriteRequest write_request;
auto add_one_entry = [&write_request](const std::string &key_string) {
    auto entity = write_request.add_entity();
    // ...
};
add_one_entry(std::string("\x00\x00\x00\x0a", 4));
add_one_entry(std::string("\x00\x00\x00\x05", 4));
```

Reply:
Fixed. Thanks for the review comments. It looks much better now.